secant/Features/BackupFlow/RecoveryPhraseValidation.swift

```diff
@@ -0,0 +1,229 @@
+//
+//  RecoveryPhraseValidation.swift
+//  secant-testnet
+//
+//  Created by Francisco Gindre on 10/29/21.
+//
+
+import Foundation
+import ComposableArchitecture
+
+typealias RecoveryPhraseValidationStore = Store<RecoveryPhraseValidationState, RecoveryPhraseValidationAction>
+typealias RecoveryPhraseValidationViewStore = ViewStore<RecoveryPhraseValidationState, RecoveryPhraseValidationAction>
+
+struct RecoveryPhraseEnvironment {
+    var mainQueue: AnySchedulerOf<DispatchQueue>
+    var newPhrase: () -> Effect<RecoveryPhrase, AppError>
+}
+
+/// Represents the completion of a group of recovery words by de addition of one word into the given group
+struct RecoveryPhraseStepCompletion: Equatable {
+    var groupIndex: Int
+    var word: String
+}
+
+struct RecoveryPhraseValidationState: Equatable {
+    enum Step: Equatable {
+        case initial
+        case incomplete
+        case complete
+        case valid
+        case invalid
+    }
+
+    static let wordGroupSize = 6
+    static let phraseChunks = 4
+
+    var phrase: RecoveryPhrase
+    var missingIndices: [Int]
+    var missingWordChips: [PhraseChip.Kind]
+    var completion: [RecoveryPhraseStepCompletion]
+
+    var step: Step {
+        guard !completion.isEmpty else {
+            return  .initial
+        }
+
+        guard completion.count >= missingIndices.count else {
+            return .incomplete
+        }
+
+        return .complete
+    }
+
+    var isValid: Bool {
+        Self.resultingPhrase(from: completion, missingIndices: missingIndices, originalPhrase: phrase, numberOfGroups: missingIndices.count) == phrase.words
+    }
+}
+
+
+extension RecoveryPhraseValidationState {
+    static func initial(
+        phrase: RecoveryPhrase,
+        missingIndices: [Int],
+        missingWordsChips: [PhraseChip.Kind]
+    ) -> Self {
+        RecoveryPhraseValidationState(
+            phrase: phrase,
+            missingIndices: missingIndices,
+            missingWordChips: missingWordsChips,
+            completion: []
+        )
+    }
+
+    /// creates an initial `RecoveryPhraseValidationState` with no completions and random missing indices.
+    /// - Note: Use this function to create a random validation puzzle for a given phrase.
+    static func random(phrase: RecoveryPhrase) -> Self {
```

Comment:
Instead of a static function, this could just be an initializer if I'm understanding this correctly.

Reply:
I think it could be confusing to the reader. There's a substancial different between the member-wise initializer and the resulting random initializer that semantically will look like a false-friend, something that in appearance does something but it's quite the opposite in reality.

````Swift
extension RecoveryPhraseValidationState {
    init(phrase: RecoveryPhrase) {
         let missingIndices = Self.randomIndices()
        let missingWordChipKind = Self.pickWordsFromMissingIndices(indices: missingIndices, phrase: phrase).shuffled()
        return RecoveryPhraseValidationState(
            phrase: phrase,
            missingIndices: missingIndices,
            missingWordChips: missingWordChipKind,
            completion: []
        )
    }
}
````

will be innocently autocompleted as RecoveryPhraseValidationState(phrase: ... )` but there's more to it that just a convenience method. That's why I chose to be explicit about it.